lib/tsconfig-parser.js

```diff
@@ -0,0 +1,139 @@
+  code: '',
+  filePath: ''
+};
+
+/**
+ * Appropriately report issues found when reading a config file
+ * @param {ts.Diagnostic} diagnostic The diagnostic raised when creating a program
+ * @returns {void}
+ */
+function diagnosticReporter(diagnostic) {
+  throw new Error(
+    ts.flattenDiagnosticMessageText(diagnostic.messageText, ts.sys.newLine)
+  );
+}
+
+const noopFileWatcher = { close: () => {} };
+
+/**
+ * Calculate project environments using options provided by eslint and paths from config
+ * @param {string} code The code being linted
+ * @param {Object} options Options provided by ESLint core
+ * @param {string} options.cwd The current working directory for the eslint process
+ * @param {string} options.filePath The path of the file being parsed
+ * @param {Object} extra Validated parser options
+ * @param {string[]} extra.project Provided tsconfig paths
+ * @returns {ts.Program[]} The programs corresponding to the supplied tsconfig paths
+ */
+module.exports = function calculateProjectParserOptions(code, options, extra) {
+  const results = [];
+  const cwd = options.cwd || process.cwd();
+
+  // preserve reference to code and file being linted
+  currentLintOperationState.code = code;
+  currentLintOperationState.filePath = options.filePath;
+
+  // Update file version if necessary
+  // TODO: only update when necessary, currently marks as changed on every lint
+  const watchCallback = watchCallbackTrackingMap.get(options.filePath);
+  if (typeof watchCallback !== 'undefined') {
+    watchCallback(options.filePath, ts.FileWatcherEventKind.Changed);
+  }
+
+  for (let tsconfigPath of extra.project) {
+    // if absolute paths aren't provided, make relative to cwd
+    if (!path.isAbsolute(tsconfigPath)) {
+      tsconfigPath = path.join(cwd, tsconfigPath);
+    }
+
+    const existingWatch = knownWatchProgramMap.get(tsconfigPath);
+
+    if (typeof existingWatch !== 'undefined') {
+      // get new program (updated if necessary)
+      results.push(existingWatch.getProgram().getProgram());
+      continue;
+    }
+
+    // create compiler host
+    const watchCompilerHost = ts.createWatchCompilerHost(
+      tsconfigPath,
+      /*optionsToExtend*/ undefined,
+      ts.sys,
+      ts.createSemanticDiagnosticsBuilderProgram,
+      diagnosticReporter,
+      /*reportWatchStatus*/ () => {}
+    );
+
+    // ensure readFile reads the code being linted instead of the copy on disk
+    const oldReadFile = watchCompilerHost.readFile;
+    watchCompilerHost.readFile = (filePath, encoding) =>
+      path.normalize(filePath) ===
+      path.normalize(currentLintOperationState.filePath)
+        ? currentLintOperationState.code
+        : oldReadFile(filePath, encoding);
+
+    // ensure process reports error on failure instead of exiting process immediately
+    watchCompilerHost.onUnRecoverableConfigFileDiagnostic = diagnosticReporter;
+
+    // ensure process doesn't emit programs
+    watchCompilerHost.afterProgramCreate = undefined;
```

Comment:
I think instead of this you just want to report errors in the program? (Because normally this is where we get diagnostics and emit

Reply:
Because I see below you are reporting  `program.getConfigFileParsingDiagnostics()` you could do that here instead so that it updates the errors as the config file updates.

Reply:
This tool doesn't actually care about errors that we determine exist in the program; that's up to the editor that the user might have. The idea here is to just provide the program to lint rule authors so that they can do what they want with it.

Reply:
Then you should use createAbstractBuilder (that just creates program nothing else)

Reply:
Moved config file error reporting to `afterProgramCreate` as you suggested; seems like a normal `BuilderProgram` doesn't work with that, so I kept it as a `SemanticDiagnosticsBuilderProgram`.